src/archive_librarian/dbHandler.cpp

```diff
@@ -0,0 +1,1456 @@
+
+    if (jobId == -1) {
+        printf("[ERROR] JobRecord was not preceded by Spawn ad, no matching JobId\n"
+                "Writing JobAd for ClusterId %d, ProcId %d\n", 
+                jobRecord.ClusterId, jobRecord.ProcId);
+        insertUnseenJob(jobRecord.Owner, jobRecord.ClusterId, jobRecord.ProcId, jobRecord.CompletionDate);
+
+        // Lookup freshly inserted info
+        jobIdCacheResults = jobIdLookup(jobRecord.ClusterId, jobRecord.ProcId);
+        jobId = jobIdCacheResults.first;
+        jobListId = jobIdCacheResults.second;
+    }
+
+    const char* sql = R"(
+        INSERT INTO JobRecords (Offset, CompletionDate, JobId, FileId, JobListId)
+        VALUES (?, ?, ?, ?, ?);
+    )";
+
+    sqlite3_stmt* stmt = nullptr;
+    if (sqlite3_prepare_v2(db_, sql, -1, &stmt, nullptr) == SQLITE_OK) {
+        sqlite3_bind_int64(stmt, 1, jobRecord.Offset);
+        sqlite3_bind_int64(stmt, 2, jobRecord.CompletionDate);
+        sqlite3_bind_int(stmt, 3, jobId);
+        sqlite3_bind_int(stmt, 4, jobRecord.FileId);
+        sqlite3_bind_int(stmt, 5, jobListId);
+
+        if (sqlite3_step(stmt) != SQLITE_DONE) {
+            printf("[ERROR] Failed to insert into JobRecords: %s\n", sqlite3_errmsg(db_));
+        }
+    } else {
+            printf("[ERROR] Failed to prepare JobRecords insert: %s \n", sqlite3_errmsg(db_));
+    }
+
+    if (stmt) sqlite3_finalize(stmt);
+}
+
+/**
+ * Batch insert multiple JobRecords in a single transaction.
+ */
+void DBHandler::batchInsertJobRecords(const std::vector<JobRecord>& jobs) {
+    if (jobs.empty()) {
+        return;
+    }
+
+    const char* sql = R"(
+        INSERT INTO JobRecords (Offset, CompletionDate, JobId, FileId, JobListId)
+        VALUES (?, ?, ?, ?, ?);
+    )";
+
+    sqlite3_stmt* stmt = nullptr;
+    int rc = sqlite3_prepare_v2(db_, sql, -1, &stmt, nullptr);
+    if (rc != SQLITE_OK) {
+        printf("[ERROR] Failed to prepare JobRecords insert: %s \n", sqlite3_errmsg(db_));
+        return;
+    }
+
+    for (const auto& job : jobs) {
+        std::pair <int, int> jobIdCacheResults = jobIdLookup(job.ClusterId, job.ProcId);
+        int jobId = jobIdCacheResults.first;
+        int jobListId = jobIdCacheResults.second;
+
+        if (jobId == -1) { 
+            printf("[ERROR] JobRecord was not preceded by Spawn ad, no matching JobId\n"
+                "Writing JobAd for ClusterId %d, ProcId %d\n", 
+                job.ClusterId, job.ProcId);
+
+            bool insertSuccess = insertUnseenJob(job.Owner, job.ClusterId, job.ProcId, job.CompletionDate);
+            if(!insertSuccess){
+                printf("JobAd for ClusterId %d, ProcId %d failed! Skipping ... \n", job.ClusterId, job.ProcId);
+                continue;
+            }
+            // Lookup freshly inserted info
+            jobIdCacheResults = jobIdLookup(job.ClusterId, job.ProcId);
+            jobId = jobIdCacheResults.first;
+            jobListId = jobIdCacheResults.second;
```

Comment:
Duplicate changes suggested in adding a single record (i.e. auto unpack and std::tie)